### DISCUSSION TITLE
Substructure Search for Purine

### DISCUSSION THREAD

**Discussion by syedzayyan** [rdkit/rdkit#6564] [Q&A] [answered]
Hello All!

Does anyone know how to do a substructure search for Purine?

I made up this code:

```
def purine_finder(smi):
    m = Chem.MolFromSmiles(smi)
    patt = Chem.MolFromSmiles('C1=C2C(=NC=N1)N=CN2')
    return m.HasSubstructMatch(patt)
```
<img width="254" alt="Screenshot 2023-07-20 at 10 59 50" src="https://github.com/rdkit/rdkit/assets/44922710/148978e5-b48a-4abb-a674-25b68fb25c8e">

It works if it is an exact match with that ^. But can be made more open? For example not having a hydrogen and have some sort of regex type query.

Thanks!

**Comment by rogersayle** [comment, marked as answer]
Try changing the patt definition to
`patt = Chem.MolFromSmarts('[#6X3]12-,=,:[#7]-,=,:[#6X3]-,=,:[#7]-,=,:[#6X3]-,=,:[#6X3]-,=,:1-,=,:[#7]-,=,:[#6X3]-,=,:[#7]-,=,:2')`
This may look complicated, but each bond becomes -,=,: (i.e. single or double or aromatic), each carbon becomes "[<span>#</span>6X3]" and each nitrogen becomes [<span>#</span>7].  This allows you to match (tautomeric) substructures such as those below which aren't hit by your original purine_finder. 

![image](https://github.com/rdkit/rdkit/assets/13512313/ac536736-6e4d-42e4-b80f-cf4726857bed)
![image](https://github.com/rdkit/rdkit/assets/13512313/fe03bf71-8cdb-4963-8513-a22a0fdfd4c6)

**Reply by syedzayyan** [reply]
Thanks very much for this. :)

This was very helpful!
